[WIP] Doc add packages

NeroBurner opened the pull request:
move some package descriptions from wiki-pages to sphix documentation

NeroBurner commented:
added labels to link in sectioned lists

ruslo commented on docs/packages/pkg/BZip2.rst:
```diff
@@ -0,0 +1,21 @@
+-  Available since
+   `v0.9.9 <https://github.com/ruslo/hunter/releases/tag/v0.9.9>`__
+
+.. code:: cmake
```
It should be `.. code-block:: cmake`

NeroBurner commented:
added script and converted files. Divided into three sections
- packages which name was in hunter and wiki
- packages only mentioned in hunter
- packages only mentioned in wiki

NeroBurner commented:
the matching at the time is a bit tricky. Here are the numbers
```
pkg_match entries:        100
pkg_only_hunter entries:  115
pkg_only_wiki entries:    41
```

NeroBurner commented:
there are many spelling errors, but I think we need to manually fix those

NeroBurner commented:
ci can't build the documentation

```
Warning, treated as error:

/home/travis/build/ruslo/hunter/docs/packages/pkg/SDL2.rst:20: WARNING: Could not lex literal_block as "cmake". Highlighting skipped.
```

the referenced lines in `SLD2.rst`
```
.. code-block:: cmake

    hunter_add_package(SDL2)
    find_package(SDL2 CONFIG REQUIRED)
    ...
    target_link_libraries(... SDL2::SDL2)
```

on my local machine `make html` runs without problems. I'm on Ubuntu 16.04 amd64.
I'm using the following script to install all dependencies, build the documentation and run spell checking
```
#!/bin/bash
set -e

echo "### Update apt cache ###"
sudo apt-get update
echo "### install sphinx packages ###"
echo "  - pyhton3-sphinx           - base package"
echo "  - python3-sphinx-rtd-theme - nice read-the-docs theme"
echo "  - python3-sphinxcontrib.spelling - spellcheck sphinx files"
#echo "  - texlive-latex-extra      - latex support to render pdf and also math in html"
#echo "  - dvipng                   - convert math formulas to pngs for html output"
#echo "  - graphviz                 - support for dot, creating flowcharts and other graphs"
echo "  - python3-matplotlib       - support for matplotlib plots"
sudo apt-get -y install python3-sphinx python3-sphinx-rtd-theme python3-sphinxcontrib.spelling python3-matplotlib

echo "### remove old build directories ###"
rm -rf _build _static _spelling
echo "### create _static dir ###"
mkdir _static

echo "### run sphinx-build ###"
make html
echo "### run sphinx spell checking ###"
sphinx-build -b spelling . _spelling
```

ruslo commented:
```
    ...
    target_link_libraries(... SDL2::SDL2)
```

I guess this is because of `...` - it's invalid CMake syntax.

ruslo commented on docs/packages/all.rst:
```diff
@@ -1,112 +1,32 @@
+
+   only_hunter/*
+
+Package description from the wiki
```
I guess the purprose of the script is to make this section empty?

NeroBurner replied:
I have three cases in the script:
- pkg: found `pkg.<pkg_name>.md` file, found corresponding `cmake/projects/<pkg_name>` folder
- only_hunter: found a hunter folder, but no corresponding markdown wiki file
    - create a template, because we have no wiki entry
- only_wiki: found a markdown wiki file, but no corresponding hunter folder
    - add a note to the rst file, that no hunter folder was found

ruslo replied:
> add a note to the rst file, that no hunter folder was found

It's a error

ruslo commented on docs/packages/all.rst:
```diff
@@ -1,112 +1,32 @@
+
+   pkg/*
+
+Packages without description
```
I don't think this one needed.

ruslo replied:
I think it make sense to add all templates to main list. They can be fixed one-by-one later without moving from one folder to another.

ruslo commented on docs/packages/only_hunter/AllTheFlopsThreads.rst:
```diff
@@ -0,0 +1,10 @@
+AllTheFlopsThreads
+==================
+
+TODO
```
Template can be helpful here, like:

```
-  `__FIXME__ Official <https://__FIXME__>`__
-  `__FIXME__ Hunterized <https://github.com/hunter-packages/__FIXME__>`__
-  `__FIXME__ Example <https://github.com/ruslo/hunter/blob/master/examples/__FIXME__/CMakeLists.txt>`__
-  Available since `__FIXME__ vX.Y.Z <https://github.com/ruslo/hunter/releases/tag/vX.Y.Z>`__
-  Added by `__FIXME__ <https://github.com/__FIXME__>`__ (`__FIXME__ pr-N <https://github.com/ruslo/hunter/pull/N>`__)

.. code-block:: cmake

    hunter_add_package(__FIXME__)
    find_package(__FIXME__ CONFIG REQUIRED)
    target_link_libraries(foo __FIXME__::__FIXME__)
```

ruslo replied:
> Available since `__FIXME__ vX.Y.Z <https://github.com/ruslo/hunter/releases/tag/vX.Y.Z>`__

We can use badge here:
* https://github.com/ruslo/hunter/issues/513

ruslo commented on docs/packages/only_wiki/ARM_NEON_2_x86_SSE.rst:
```diff
@@ -0,0 +1,24 @@
+==================
+
+File only available in wiki.
+Please merge with hunter-package entry
```
What does it mean?

NeroBurner replied:
for this particular package it means, that my matching algorithm is wrong. For other packages it means, that for `pkg.<pkg_name>.md` there was no `cmake/projects/<pkg_name>` folder.

The markdown filename is assumed to be lower case, thats why `ARM_NEON_2_X86_SSE` is not found. For others the markdown filename and the hunter-folder name don't match

ruslo replied:
> that for pkg.<pkg_name>.md there was no cmake/projects/<pkg_name> folder

Can you create the list with such packages, I will take a look.

NeroBurner replied:
the full list of packages with no documentation vs wiki entries with no matching hunter-projects

```
only_hunter/AllTheFlopsThreads.rst
only_hunter/Clang.rst
only_hunter/ClangToolsExtra.rst
only_hunter/CLAPACK.rst
only_hunter/clBLAS.rst
only_hunter/CppNetlib.rst
only_hunter/crashup.rst
only_hunter/czmq.rst
only_hunter/damageproto.rst
only_hunter/dbus.rst
only_hunter/dri2proto.rst
only_hunter/dri3proto.rst
only_hunter/drishti_assets.rst
only_hunter/drishti_faces.rst
only_hunter/drm.rst
only_hunter/eigen3-nnls.rst
only_hunter/enet.rst
only_hunter/fixesproto.rst
only_hunter/geos.rst
only_hunter/getopt.rst
only_hunter/glew.rst
only_hunter/glproto.rst
only_hunter/ice.rst
only_hunter/Igloo.rst
only_hunter/inputproto.rst
only_hunter/intltool.rst
only_hunter/ios_sim.rst
only_hunter/ippicv.rst
only_hunter/kbproto.rst
only_hunter/LAPACK.rst
only_hunter/Leptonica.rst
only_hunter/Libcxx.rst
only_hunter/Libcxxabi.rst
only_hunter/libdaemon.rst
only_hunter/libjson-rpc-cpp.rst
only_hunter/libogg.rst
only_hunter/librtmp.rst
only_hunter/libxml2.rst
only_hunter/LLVMCompilerRT.rst
only_hunter/mini_chromium.rst
only_hunter/mpark_variant.rst
only_hunter/mtplz.rst
only_hunter/MySQL-client.rst
only_hunter/nanoflann.rst
only_hunter/NASM.rst
only_hunter/odb.rst
only_hunter/odb-boost.rst
only_hunter/odb-compiler.rst
only_hunter/odb-mysql.rst
only_hunter/odb-sqlite.rst
only_hunter/onmt.rst
only_hunter/OpenCV-Extra.rst
only_hunter/OpenNMTTokenizer.rst
only_hunter/pciaccess.rst
only_hunter/presentproto.rst
only_hunter/pthread-stubs.rst
only_hunter/pugixml.rst
only_hunter/QtCMakeExtra.rst
only_hunter/randrproto.rst
only_hunter/renderproto.rst
only_hunter/SDL_mixer.rst
only_hunter/sm.rst
only_hunter/Snappy.rst
only_hunter/Sober.rst
only_hunter/Sources-for-Android-SDK.rst
only_hunter/SQLite3.rst
only_hunter/SuiteSparse.rst
only_hunter/Tesseract.rst
only_hunter/tomcrypt.rst
only_hunter/tommath.rst
only_hunter/Washer.rst
only_hunter/WinSparkle.rst
only_hunter/x11.rst
only_hunter/x264.rst
only_hunter/xau.rst
only_hunter/xcb.rst
only_hunter/xcb-proto.rst
only_hunter/xcursor.rst
only_hunter/xdamage.rst
only_hunter/xext.rst
only_hunter/xextproto.rst
only_hunter/xf86vidmodeproto.rst
only_hunter/xfixes.rst
only_hunter/xi.rst
only_hunter/xinerama.rst
only_hunter/xineramaproto.rst
only_hunter/xorg-macros.rst
only_hunter/xproto.rst
only_hunter/xrandr.rst
only_hunter/xrender.rst
only_hunter/xshmfence.rst
only_hunter/xtrans.rst
only_hunter/xxf86vm.rst
only_hunter/zookeeper.rst
```

vs

```
only_wiki/accelerate.rst
only_wiki/appkit.rst
only_wiki/assetslibrary.rst
only_wiki/audiotoolbox.rst
only_wiki/avfoundation.rst
only_wiki/coredata.rst
only_wiki/corefoundation.rst
only_wiki/coregraphics.rst
only_wiki/corelocation.rst
only_wiki/coremedia.rst
only_wiki/corevideo.rst
only_wiki/egl.rst
only_wiki/foundation.rst
only_wiki/glapi.rst
only_wiki/gles2.rst
only_wiki/gles3.rst
only_wiki/glkit.rst
only_wiki/imageio.rst
only_wiki/iokit.rst
only_wiki/metal.rst
only_wiki/mobilecoreservices.rst
only_wiki/mysql.client.rst
only_wiki/opengles.rst
only_wiki/osmesa.rst
only_wiki/quartzcore.rst
only_wiki/uikit.rst
```

NeroBurner commented:
fixed all the rst files in `packages/pkg`

NeroBurner commented:
there are packages for `LLVM` and `Clang`. In the wiki there is only `pkg.llvm.clang.rst`. Which documentation is it?

in the usage it is
```
hunter_add_package(LLVM)
```

so I will add it to the LLVM documentation

I'm also lost at the Android documentation, which wiki pages correspond to which hunter packages

NeroBurner commented:
I've merged all `only_wiki` entries to `only_hunter` entries I could find. For remaining files in `only_wiki` I don't know which packages they describe

At the end of this PR the folder `only_wiki` should be empty

ruslo commented:
> In the wiki there is only pkg.llvm.clang.rst. Which documentation is it?

`pkg.llvm` I've fixed it.

> I'm also lost at the Android documentation, which wiki pages correspond to which hunter packages

Names should match. Can you give an example?

NeroBurner commented:
```
only_hunter/Android-Apk.rst
only_hunter/Android-ARM-EABI-v7a-System-Image.rst
only_hunter/Android-Build-Tools.rst
only_hunter/Android-Google-APIs.rst
only_hunter/Android-Google-APIs-Intel-x86-Atom-System-Image.rst
only_hunter/Android-Google-Repository.rst
only_hunter/Android-Intel-x86-Atom-System-Image.rst
only_hunter/Android-Modules.rst
only_hunter/Android-SDK.rst
only_hunter/Android-SDK-Platform.rst
only_hunter/Android-SDK-Platform-tools.rst
only_hunter/Android-SDK-Tools.rst
only_hunter/Android-Support-Repository.rst
```

vs

```
only_wiki/android.apk.rst
only_wiki/android.modules.rst
only_wiki/android.rst
only_wiki/android.sdk.rst
only_wiki/android_log.rst
```

which packages do I combine?

ruslo commented on docs/packages/only_hunter/AllTheFlopsThreads.rst:
```diff
@@ -0,0 +1,21 @@
+
+AllTheFlopsThreads
+==================
+
```
Looks good, another enchancement is to add warning:
```
.. warning::

  This page is a template and contains no real information.
  Please send pull request with real description.
```

ruslo commented on docs/packages/only_wiki/android_log.rst:
```diff
@@ -0,0 +1,16 @@
+android_log
+===========
+
+File only available in wiki.
```
It's a helper package, there will be no corresponding `hunter_add_package`.

List (created by `grep -IR hunter_find_helper cmake/find`):
* accelerate
* android
* android_log
* appkit
* assetslibrary
* audiotoolbox
* avfoundation
* coredata
* corefoundation
* coregraphics
* corelocation
* coremedia
* corevideo
* egl
* foundation
* glapi
* gles2
* gles3
* glkit
* imageio
* iokit
* metal
* mobilecoreservices
* opengles
* osmesa
* quartzcore
* uikit

NeroBurner replied:
what shall we do with helpers? Should we add them in their own folder, because they are no hunter package?

ruslo replied:
> Should we add them in their own folder

No, no need. The fact that they are special reflected by absense of `hunter_add_package` in example. We can add `.. note` section though.

NeroBurner replied:
How about this note?
```
.. note::

    This is a helper package. There is no corresponding package in hunter to be included by ``hunter_add_package(...)``
```

ruslo commented on docs/packages/only_wiki/glkit.rst:
```diff
@@ -0,0 +1,24 @@
+.. code-block:: cmake
+
+    find_package(glkit REQUIRED)
+    target_link_libraries(... glkit::glkit)
```
I think pattern `target_link_libraries(...` can be substituted with `target_link_libraries(foo` so we don't have warning about incorrect syntax.

NeroBurner replied:
the incorrect syntax was because of the line only containing `...`. the other lines compile fine (the current PR compiles the html fine, but has spelling errors

ruslo commented:
> only_hunter/Android-SDK.rst

> only_wiki/android.sdk.rst

Okay I see, I think we can't use `-` in name of wiki page. Let the rule be:
* uppercase -> lowercase
* `-` -> `_`

> only_wiki/android.apk.rst
> only_wiki/android.modules.rst
> only_wiki/android.sdk.rst

Fixed

> only_wiki/android.rst
> only_wiki/android_log.rst

It's a helper, see my notes above.

NeroBurner commented:
move android packges and mysql.client packge to `pkg`

added note for helper packages. Now the folder `only_wiki`contains just helper packages. If you accept the note (or give a more accurate note) I'll move all files from `only_wiki` to `pkg`

The same for `only_hunter`. If you are happy with the template I'll move all files from `only_hunter` to `pkg`

Then the last thing left to do is fix the spelling warnings

ruslo commented on docs/packages/compression.rst:
```diff
@@ -1,8 +1,8 @@
 
- * `BZip2 <https://github.com/ruslo/hunter/wiki/pkg.bzip2>`_ - high-quality data compressor.
+ * :ref:`pkg.BZip2` - high-quality data compressor.
  * `minizip <https://github.com/ruslo/hunter/wiki/pkg.minizip>`_ - enables to extract files from a .zip archive file.
```
Should it be
```
:ref:`pkg.minizip`
```
now?

NeroBurner replied:
Yes. The labels are created to have exactly the name the folder in `cmake/projects` has. When having `cmake/projects/LibFoo` I've generated a label `pkg.LibFoo`

ruslo commented on docs/packages/only_wiki/android.rst:
```diff
@@ -0,0 +1,17 @@
+
+.. note::
+
+    This is a helper package. There is no corresponding package in hunter to be included by ``hunter_add_package(...)``
```
`s,hunter,Hunter,`

NeroBurner commented:
I've addad a new section system_library_finder

I've also added the `.. index::` directive to all packages in that new section. The result are the packages listet in `genindex.html`

![screenshot-2017-8-30 index hunter 0 14 documentation](https://user-images.githubusercontent.com/9076163/29859794-0eb83b50-8d64-11e7-801b-8099ace0270f.png)

NeroBurner commented:
updated all sections and added an index command to all files. Files with no index I've added to 'unsorted`. With this anyone can go through the packages and sort into sections

ruslo commented:
Looks great, thank you! Will remove `pkg.*` pages from wiki and will take a look at spelling issues.

NeroBurner commented:
I just pushed spelling fixes, didn't see that you merged

NeroBurner commented:
opened a new pr with the spelling fixes
https://github.com/ruslo/hunter/pull/995

ruslo commented:
> I would add a deprecated warning to the wiki pages and maybe a link to the corresponding documentation. Older hunter version may still link to the wiki pages

I don't think it make sense, quite confusing. I'm planning to close the GitHub wiki.

ruslo commented on docs/packages/pkg/AllTheFlopsThreads.rst:
```diff
@@ -0,0 +1,28 @@
+-  `__FIXME__ Official <https://__FIXME__>`__
+-  `__FIXME__ Hunterized <https://github.com/hunter-packages/__FIXME__>`__
+-  `__FIXME__ Example <https://github.com/ruslo/hunter/blob/master/examples/__FIXME__/CMakeLists.txt>`__
+-  Available since `__FIXME__ vX.Y.Z <https://github.com/ruslo/hunter/releases/tag/vX.Y.Z>`__
```
It will be tricky to maintain this on practice. I expect this part contributed by pull request but at time when pull request is opened we don't know when package will be released exactly... I vote for not using this in new packages, user can deduce Hunter version from related pull request (see line below).

NeroBurner replied:
I personally like the "Available since" tag. It is MUCH easier for the user this way. The documentation can be updated after the release. By putting `..` before the line the whole line can be made into a comment and can be updated afterwards

ruslo replied:
> It is MUCH easier for the user this way. The documentation can be updated after the release.

I understand that but I don't know how to keep it consistent on practice.

NeroBurner replied:
yes, thats tricky. But not more tricky than keeping the wiki consistent. I think it is the same level of trickyness

I could update the template to have the "Available since" line added as comment. Then it would still be there, but not hindering contributions

ruslo replied:
> But not more tricky than keeping the wiki consistent. I think it is the same level of trickyness

Different. Before this change I was asking developers to create/update `pkg.foo` wiki page **after** merge done and I've created release, now `pkg/foo.rst` is a **part of pull request** and release version is not known for sure. User have to create second pull request with "Available since".

NeroBurner replied:
yes, you are right. This work will most probably be on your part (or MAYBE the original contributor) :-/

ruslo replied:
I've added Hunter badge to "Available since", example:
* https://docs.hunter.sh/en/latest/packages/pkg/ArrayFire.html

However I still not sure about usefulness of this feature. I see few scenarios:
* User already has URL/SHA1 in project and want to check if it's possible to add new `hunter_add_package(foo)`. He goes to `pkg.foo` and see that this package is not available for Hunter version he is currently using. He can use version from `Available since X.Y.Z` **but** it's better to update to **latest** version of Hunter since there may be new stable versions of `foo` there.
* Even if `Available since` satisfied the build of project may end with error. In this case user should use **latest** version of Hunter (at least try to confirm)
* Instead of going to `pkg.foo` he can just add `hunter_add_package(foo)` and if package is not available Hunter will report clear error about it ("project not found"). He can go and update to **latest** version of Hunter.
* User starting project from scratch and want to use project `foo`. He should start with **latest** Hunter version and there is no need to check `Available since` part.

> edit: updated genindex view

After [updating requirements.txt](https://github.com/ruslo/hunter/commit/4ee10f9b0bf1f89a25fe1abb6d8fce7e64df92cf) this page looks different:
* https://docs.hunter.sh/en/latest/genindex.html

NeroBurner replied:
yes, it's always best to update to the newest version. But it may also be useful to know how far back one can go to still have project `foo` available. I think it is not mandatory, but nice to have

yes, it looks different. A more flat style. I liked the old style better, the "subsections" were visibly more clearly.
Was there a reason for upgrade? did the style change in this version-jump? Should we somehow make our own theme for the `genindex` page?

ruslo replied:
> I liked the old style better

Me too, may be there is a way to use old style?

> Was there a reason for upgrade?

Having internal errors after badge code added I made an update and it fixed everything.

> Should we somehow make our own theme for the genindex page?

NeroBurner replied:
> Me too, may be there is a way to use old style?

maybe if the you keep `-sphinx-rtd-theme==0.1.9` in `requirements.txt`

> Should we somehow make our own theme for the genindex page?

I hoped you had more experience with theming. I was introduced to Shinx through your Hunter-Project :) And I am a C++ programmer (like you, I suspect) and don't have much clue about CSS, HTML etc

ruslo replied:
> maybe if the you keep -sphinx-rtd-theme==0.1.9 in requirements.txt

Well probably, but I don't want to move back. We will have to use new version if there will be other improvements/features introduced.

> I hoped you had more experience with theming

Nope. Only trivial copy/paste + substitute :)

NeroBurner replied:
then I think we should open an Issue and hope maybe someone dislikes the genindex page enough to fix the style ^^

ruslo commented:
> I liked the old style better

Me too, may be there is a way to use old style?

> Was there a reason for upgrade?

Having internal errors after badge code added I made an update and it fixed everything.

> Should we somehow make our own theme for the genindex page?

Not experienced enough to answer this question, pull request welcome as usual :)
